### PR TITLE
Optimise DCAR ad insertion for articles with many images

### PR DESCRIPTION
**Pull request by emma-imber**
## What does this change?
- Changes the DCAR insertion logic to insert a placeholder before the current element instead of afterwards. This aligns the code with the way that ad insertion is done for web pages. We change the minimum paragraph to consider for insertion to 4, so that 3 paragraphs still appear before the first ad, for no visible change in page experience.
- Only insert ads after images, not immediately before them. Again, this is for alignment with the current ad logic on web.
- Updates the DCAR ad insertion logic to count image elements when considering ad placement. This means that previously, you could have alternating images and paragraphs so that an ad would be shown after a total of 12 elements. Now, an ad would be shown after a minimum of 6 total elements, as long as it doesn't appear immediately before an image.
- Updates tests to account for logic change, and adds a test to check we don't insert ads after an element which isn't an image or text element.

These changes result in no difference in how ad placeholders are positioned for articles without images. See example listed below.

## Why?
On DCAR we were seeing ads inserted before images but never after, and we saw the opposite behaviour on web. The change to insert ads before the current element instead of afterwards resolves this discrepancy for consistency between platforms.

The majority of images are at least as large as a normal paragraph, and on articles with a lot of images this results in very few ads. By altering this logic to count image elements, we can improve ad ratio for these articles. The difference to page experience isn't very noticeable as these articles still have a good distance between ads. Some examples are shown below.

### Article with no images - before changes
https://github.com/guardian/dotcom-rendering/assets/108270776/bc389f91-59be-4c80-aa69-a790a95fbd2e

### Article with no images - after changes
No observable difference in the positioning or frequency of ad placeholders

https://github.com/guardian/dotcom-rendering/assets/108270776/4f297bfa-3bc1-459d-9889-8a51ac97fa82

### Ranked article - before changes
6 ads in total:

https://github.com/guardian/dotcom-rendering/assets/108270776/9cf6f5c3-6052-460f-accc-f69de8cd890f

### Ranked article - after changes
8 ads in total:

https://github.com/guardian/dotcom-rendering/assets/108270776/e867d7c5-91fa-4944-9a75-e08216f31395

### Travel article - before changes
5 ads in total:

https://github.com/guardian/dotcom-rendering/assets/108270776/e6e59381-c549-4fae-b64b-8bb82df106ca

### Travel article - after changes
7 ads in total:

https://github.com/guardian/dotcom-rendering/assets/108270776/fe184cc3-4411-4d2c-95d9-5868d4c540db

### Before changes - ads appear before images
Screenshot 2024-03-04 at 12 26 29

### After changes - ads appear after images
Screenshot 2024-03-04 at 12 26 51

### CODE REVIEW

File: dotcom-rendering/src/model/enhance-ad-placeholders.ts [hunk 1]
```diff
@@ -65,13 +83,16 @@ const insertAdPlaceholders = (elements: FEElement[]): FEElement[] => {
 			currentElement: FEElement,
 			idx: number,
 		): ReducerAccumulator => {
-			const paragraphCounter = isParagraph(currentElement)
-				? prev.paragraphCounter + 1
-				: prev.paragraphCounter;
+			const blockCounter =
```
Comment (Jakeii) [inline dotcom-rendering/src/model/enhance-ad-placeholders.ts:86]: I know it's not your addition but why only count paragraphs and now images too? And not embeds or whatever else

File: dotcom-rendering/src/model/enhance-ad-placeholders.ts [hunk 2]
```diff
@@ -31,28 +34,43 @@ const isSuitablePosition = (
...
 	};
-	return [...elements, placeholder];
+	return [...prevElements, placeholder, currentElement];
 };
 
 type ReducerAccumulator = {
 	elements: FEElement[];
-	paragraphCounter: number;
+	blockCounter: number;
+	lastAdIndex: number;
```
Comment (domlander) [inline dotcom-rendering/src/model/enhance-ad-placeholders.ts:71]: Is `lastAdIndex` the index of the most recent paragraph an ad has been inserted before? e.g. in the case that an ad has inserted before the 4th and 10th element, `lastAdIndex` would be 10? Maybe a comment would be useful here?

File: dotcom-rendering/src/model/enhance-ad-placeholders.ts [hunk 3]
```diff
@@ -5,20 +5,23 @@ import type { RenderingTarget } from '../types/renderingTarget';
 /**
  * Positioning rules:
  *
- * - Is further than 3 paragraphs in
- * - Is every 6 paragraphs
+ * - Is further than 3 blocks in
```
Comment (domlander) [inline dotcom-rendering/src/model/enhance-ad-placeholders.ts:8]: Using `elements` instead of `blocks` would be more consistent with the logic used in liveblog pages. In liveblogs, an `element` is used for some content of the type `model.dotcomrendering.pageElements.???` whereas `blocks` is used to define a group of `elements` that make up a post

File: dotcom-rendering/src/model/enhance-ad-placeholders.ts [hunk 4]
```diff
@@ -65,13 +83,16 @@ const insertAdPlaceholders = (elements: FEElement[]): FEElement[] => {
 			currentElement: FEElement,
 			idx: number,
 		): ReducerAccumulator => {
```
Comment (Jakeii) [inline dotcom-rendering/src/model/enhance-ad-placeholders.ts:85]: Not blocking, although the `reduce` works well, it's a bit hard to follow, I think it might be better to refactor to a simple loop with the accumulator in the parent scope?

**Review by domlander — APPROVED** [review]

File: dotcom-rendering/src/model/enhance-ad-placeholders.ts (hunk 1 above)
Comment (emma-imber) [reply dotcom-rendering/src/model/enhance-ad-placeholders.ts:86]: I'm assuming for simplicity? Because we're looking at the server side page elements before they're rendered, we could check a long list of element types and basically end up with an apps version of spacefinder. Images seem sensible to include as there's so many of them! Perhaps we should have a think about what we want to support and where we draw the line to avoid the code getting super complicated?

File: dotcom-rendering/src/model/enhance-ad-placeholders.ts (hunk 2 above)
Comment (emma-imber) [reply dotcom-rendering/src/model/enhance-ad-placeholders.ts:71]: Yes that's exactly it. We're using it to count that we have at least 6 blocks between ads. I can add a comment to make this clearer

File: dotcom-rendering/src/model/enhance-ad-placeholders.ts (hunk 3 above)
Comment (emma-imber) [reply dotcom-rendering/src/model/enhance-ad-placeholders.ts:8]: I can switch to using elements 👍

File: dotcom-rendering/src/model/enhance-ad-placeholders.ts (hunk 4 above)
Comment (emma-imber) [reply dotcom-rendering/src/model/enhance-ad-placeholders.ts:93]: Yeah that might be easier to read! Although IMO that isn't really in the scope of this ticket and I feel like there are enough changes in here already. Maybe we could add this idea to the backlog?

Comment (arelra) [reply dotcom-rendering/src/model/enhance-ad-placeholders.ts:93]: > a simple loop with the accumulator in the parent scope

I think this style is discouraged in DCR in favour of a more functional style.
A previous discussion:
https://github.com/guardian/dotcom-rendering/pull/8118/files

**Review by arelra — APPROVED** [review]
🚀

File: dotcom-rendering/src/model/enhance-ad-placeholders.ts (hunk 4 above)
Comment (emma-imber) [reply dotcom-rendering/src/model/enhance-ad-placeholders.ts:93]: That's really useful context, thank you for flagging this Ravi!
